src/main/java/net/minecraftforge/recipe/RecipeHandler.java

```diff
@@ -0,0 +1,244 @@
+    {
+        if (trie.possibleNextIngredients().size() == 0)
+        {
+            if (index < ingredientList.length) return null;
+            else return Pair.of(trie.getResult(), trie.getStochasticResults());
+        }
+        if (index >= ingredientList.length) return null;
+
+        AbstractIngredient nextIngredient = ingredientList[index];
+        ArrayList<RecipeNode> possiblePaths = new ArrayList<RecipeNode>();
+
+        for (RecipeNode n : trie.possibleNextIngredients()) if (n.fulfilled(nextIngredient)) possiblePaths.add(n);
+
+        if (possiblePaths.size() == 0) return null;
+
+        for (RecipeNode n : possiblePaths)
+        {
+            Pair<ItemStack, ArrayList<Pair<ItemStack, Float>>> result = recursiveTrieLookup(ingredientList, n, index + 1);
+            if (result != null) return result;
+        }
+
+        return null;
+    }
+
+    private void addRecipeToTrie(RecipeNode trie, AbstractIngredient[] ingredientList, ItemStack result, ArrayList<Pair<ItemStack, Float>> stochasticResults)
+    {
+        RecipeNode currentNode = trie;
+        for (int i = 0; i < ingredientList.length; i++)
+        {
+            AbstractIngredient nextIngredient = ingredientList[i];
+            int nextIngredientIndex = currentNode.nextIngredientIndex(nextIngredient);
+            if (nextIngredientIndex != -1) currentNode = currentNode.possibleNextIngredients().get(nextIngredientIndex);
+            else {
+                RecipeNode nextNode = new RecipeNode(nextIngredient);
+                currentNode.addNextIngredient(nextNode);
+                currentNode = nextNode;
+            }
+        }
+        
+        currentNode.setResult(result);
+        currentNode.setStochasticResults(stochasticResults);
+    }
+
+    private void addReverseLookup(HashMap<Item, ArrayList<Pair<AbstractIngredient[], Float>>> reverseLookupTable, AbstractIngredient[] ingredientList, float chance, ItemStack result)
+    {
+        Item resultItem = result.getItem();
+        if (reverseLookupTable.containsKey(resultItem)) 
+        {
+            System.out.println(reverseLookupTable.get(resultItem));
+            reverseLookupTable.get(resultItem).add(Pair.of(ingredientList, chance));
+        }
+        else
+        {
+            ArrayList<Pair<AbstractIngredient[], Float>> newRecipeList = new ArrayList<Pair<AbstractIngredient[], Float>>();
+            newRecipeList.add(Pair.of(ingredientList, chance));
+            reverseLookupTable.put(resultItem, newRecipeList);
+        }
+    }
+    
+    /**
+     * This internal class represents the prefix tree structure.
+     */
+    private class RecipeNode
+    {
+        private ArrayList<RecipeNode> possibleNextIngredients;
+        private AbstractIngredient ingredient;
+        private ItemStack result;
+        private ArrayList<Pair<ItemStack, Float>> stochasticResults;
```

Comment:
Just a note, best practice is to keep this as loose as possible, so you'd want List all over the place not ArrayList.